File: backend/src/protocol.zig
const std = @import("std");

// ── Client → Server Messages ──────────────────────────────────────────

pub const NewSessionPayload = struct {};

pub const AttachPayload = struct {
    session_id: []const u8,
};

pub const InputPayload = struct {
    session_id: []const u8,
    data: []const u8,
};

pub const ResizePayload = struct {
    session_id: []const u8,
    cols: u16,
    rows: u16,
};

pub const ClientMessage = union(enum) {
    new_session: NewSessionPayload,
    attach: AttachPayload,
    input: InputPayload,
    resize: ResizePayload,
};

// ── Server → Client Messages ──────────────────────────────────────────

pub const SessionCreatedPayload = struct {
    session_id: []const u8,
};

pub const OutputPayload = struct {
    session_id: []const u8,
    data: []const u8,
};

pub const SessionEndedPayload = struct {
    session_id: []const u8,
};

pub const ErrorPayload = struct {
    message: []const u8,
};

pub const ServerMessage = union(enum) {
    session_created: SessionCreatedPayload,
    output: OutputPayload,
    session_ended: SessionEndedPayload,
    @"error": ErrorPayload,
};

// ── JSON structure for parsing ────────────────────────────────────────

const RawMessage = struct {
    type: []const u8,
    session_id: ?[]const u8 = null,
    data: ?[]const u8 = null,
    cols: ?u16 = null,
    rows: ?u16 = null,
    message: ?[]const u8 = null,
};

// ── Parsing ───────────────────────────────────────────────────────────

pub const ParseError = error{
    UnknownMessageType,
    MissingField,
};

pub fn parseClientMessage(json: []const u8) !ClientMessage {
    const parsed = std.json.parseFromSlice(RawMessage, std.heap.page_allocator, json, .{
        .ignore_unknown_fields = true,
    }) catch {
        return error.UnknownMessageType;
    };
    const raw = parsed.value;

    if (std.mem.eql(u8, raw.type, "new_session")) {
        return .{ .new_session = .{} };
    } else if (std.mem.eql(u8, raw.type, "attach")) {
        return .{ .attach = .{
            .session_id = raw.session_id orelse return error.MissingField,
        } };
    } else if (std.mem.eql(u8, raw.type, "input")) {
        return .{ .input = .{
            .session_id = raw.session_id orelse return error.MissingField,
            .data = raw.data orelse return error.MissingField,
        } };
    } else if (std.mem.eql(u8, raw.type, "resize")) {
        return .{ .resize = .{
            .session_id = raw.session_id orelse return error.MissingField,
            .cols = raw.cols orelse return error.MissingField,
            .rows = raw.rows orelse return error.MissingField,
        } };
    }

    return error.UnknownMessageType;
}

// ── Serialization ─────────────────────────────────────────────────────

pub fn serializeServerMessage(allocator: std.mem.Allocator, msg: ServerMessage) ![]const u8 {
    var buf = std.ArrayList(u8).init(allocator);
    errdefer buf.deinit();
    const writer = buf.writer();

    try writer.writeAll("{");

    switch (msg) {
        .session_created => |payload| {
            try writer.writeAll("\"type\":\"session_created\",\"session_id\":");
            try std.json.stringify(payload.session_id, .{}, writer);
        },
        .output => |payload| {
            try writer.writeAll("\"type\":\"output\",\"session_id\":");
            try std.json.stringify(payload.session_id, .{}, writer);
            try writer.writeAll(",\"data\":");
            try std.json.stringify(payload.data, .{}, writer);
        },
        .session_ended => |payload| {
            try writer.writeAll("\"type\":\"session_ended\",\"session_id\":");
            try std.json.stringify(payload.session_id, .{}, writer);
        },
        .@"error" => |payload| {
            try writer.writeAll("\"type\":\"error\",\"message\":");
            try std.json.stringify(payload.message, .{}, writer);
        },
    }

    try writer.writeAll("}");

    return buf.toOwnedSlice();
}

// ── Tests ──────────────────────────────────────────────────────────────

test "parse new_session message" {
    const msg = try parseClientMessage("{\"type\":\"new_session\"}");
    try std.testing.expect(msg == .new_session);
}

test "parse attach message" {
    const msg = try parseClientMessage("{\"type\":\"attach\",\"session_id\":\"abc123\"}");
    try std.testing.expect(msg == .attach);
    try std.testing.expectEqualStrings("abc123", msg.attach.session_id);
}

test "parse input message" {
    const msg = try parseClientMessage("{\"type\":\"input\",\"session_id\":\"abc123\",\"data\":\"ls\\r\"}");
    try std.testing.expect(msg == .input);
    try std.testing.expectEqualStrings("abc123", msg.input.session_id);
    try std.testing.expectEqualStrings("ls\r", msg.input.data);
}

test "parse resize message" {
    const msg = try parseClientMessage("{\"type\":\"resize\",\"session_id\":\"abc123\",\"cols\":120,\"rows\":40}");
    try std.testing.expect(msg == .resize);
    try std.testing.expectEqualStrings("abc123", msg.resize.session_id);
    try std.testing.expectEqual(@as(u16, 120), msg.resize.cols);
    try std.testing.expectEqual(@as(u16, 40), msg.resize.rows);
}

test "parse invalid message type returns error" {
    const result = parseClientMessage("{\"type\":\"unknown\"}");
    try std.testing.expectError(error.UnknownMessageType, result);
}

test "parse malformed json returns error" {
    const result = parseClientMessage("not json at all");
    try std.testing.expect(if (result) |_| false else |_| true);
}

test "serialize session_created message" {
    const allocator = std.testing.allocator;
    const json = try serializeServerMessage(allocator, .{
        .session_created = .{ .session_id = "abc123" },
    });
    defer allocator.free(json);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json, .{});
    defer parsed.deinit();

    const obj = parsed.value.object;
    try std.testing.expectEqualStrings("session_created", obj.get("type").?.string);
    try std.testing.expectEqualStrings("abc123", obj.get("session_id").?.string);
}

test "serialize output message" {
    const allocator = std.testing.allocator;
    const json = try serializeServerMessage(allocator, .{
        .output = .{ .session_id = "abc123", .data = "hello\n" },
    });
    defer allocator.free(json);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json, .{});
    defer parsed.deinit();

    const obj = parsed.value.object;
    try std.testing.expectEqualStrings("output", obj.get("type").?.string);
    try std.testing.expectEqualStrings("abc123", obj.get("session_id").?.string);
    try std.testing.expectEqualStrings("hello\n", obj.get("data").?.string);
}

test "serialize session_ended message" {
    const allocator = std.testing.allocator;
    const json = try serializeServerMessage(allocator, .{
        .session_ended = .{ .session_id = "abc123" },
    });
    defer allocator.free(json);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json, .{});
    defer parsed.deinit();

    const obj = parsed.value.object;
    try std.testing.expectEqualStrings("session_ended", obj.get("type").?.string);
    try std.testing.expectEqualStrings("abc123", obj.get("session_id").?.string);
}

test "serialize error message" {
    const allocator = std.testing.allocator;
    const json = try serializeServerMessage(allocator, .{
        .@"error" = .{ .message = "not found" },
    });
    defer allocator.free(json);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json, .{});
    defer parsed.deinit();

    const obj = parsed.value.object;
    try std.testing.expectEqualStrings("error", obj.get("type").?.string);
    try std.testing.expectEqualStrings("not found", obj.get("message").?.string);
}

File: backend-swift/Sources/Terminatab/Protocol.swift
import Foundation

// MARK: - Client → Server Messages

enum ClientMessage {
    case newSession
    case attach(sessionId: String)
    case input(sessionId: String, data: String)
    case resize(sessionId: String, cols: UInt16, rows: UInt16)
}

// MARK: - Server → Client Messages

enum ServerMessage: Sendable {
    case sessionCreated(sessionId: String)
    case output(sessionId: String, data: Data)
    case sessionEnded(sessionId: String)
    case error(message: String)
}

// MARK: - Parsing

enum ProtocolError: Error {
    case unknownMessageType
    case missingField
    case invalidJSON
}

private struct RawMessage: Decodable {
    let type: String
    let session_id: String?
    let data: String?
    let cols: UInt16?
    let rows: UInt16?
    let message: String?
}

func parseClientMessage(_ json: String) throws -> ClientMessage {
    guard let data = json.data(using: .utf8) else {
        throw ProtocolError.invalidJSON
    }

    let raw: RawMessage
    do {
        raw = try JSONDecoder().decode(RawMessage.self, from: data)
    } catch {
        throw ProtocolError.invalidJSON
    }

    switch raw.type {
    case "new_session":
        return .newSession
    case "attach":
        guard let sessionId = raw.session_id else { throw ProtocolError.missingField }
        return .attach(sessionId: sessionId)
    case "input":
        guard let sessionId = raw.session_id, let inputData = raw.data else {
            throw ProtocolError.missingField
        }
        return .input(sessionId: sessionId, data: inputData)
    case "resize":
        guard let sessionId = raw.session_id, let cols = raw.cols, let rows = raw.rows else {
            throw ProtocolError.missingField
        }
        return .resize(sessionId: sessionId, cols: cols, rows: rows)
    default:
        throw ProtocolError.unknownMessageType
    }
}

// MARK: - Serialization

func serializeServerMessage(_ msg: ServerMessage) -> String {
    switch msg {
    case .sessionCreated(let sessionId):
        return "{\"type\":\"session_created\",\"session_id\":\(jsonString(sessionId))}"
    case .output(let sessionId, let data):
        return String(decoding: serializeOutputData(sessionId: sessionId, data: data), as: UTF8.self)
    case .sessionEnded(let sessionId):
        return "{\"type\":\"session_ended\",\"session_id\":\(jsonString(sessionId))}"
    case .error(let message):
        return "{\"type\":\"error\",\"message\":\(jsonString(message))}"
    }
}

/// JSON-encode a Swift String value with proper escaping.
private func jsonString(_ s: String) -> String {
    var result = "\""
    for char in s {
        switch char {
        case "\"": result += "\\\""
        case "\\": result += "\\\\"
        case "\n": result += "\\n"
        case "\r": result += "\\r"
        case "\t": result += "\\t"
        default:
            if char.asciiValue != nil && char.asciiValue! < 0x20 {
                result += String(format: "\\u%04x", char.asciiValue!)
            } else {
                result.append(char)
            }
        }
    }
    result += "\""
    return result
}

/// JSON-escape raw bytes into a [UInt8] buffer (a quoted JSON string value).
/// Returns raw bytes to avoid String(decoding:as:UTF8.self) which replaces
/// incomplete UTF-8 sequences (from chunk boundaries) with U+FFFD.
private func jsonEscapeBytes(_ data: Data) -> [UInt8] {
    var buf: [UInt8] = [0x22] // opening "
    for byte in data {
        switch byte {
        case 0x22: buf.append(contentsOf: [0x5C, 0x22])       // \"
        case 0x5C: buf.append(contentsOf: [0x5C, 0x5C])       // \\
        case 0x0A: buf.append(contentsOf: [0x5C, 0x6E])       // \n
        case 0x0D: buf.append(contentsOf: [0x5C, 0x72])       // \r
        case 0x09: buf.append(contentsOf: [0x5C, 0x74])       // \t
        case 0x00...0x08, 0x0B, 0x0C, 0x0E...0x1F:
            let hex = String(format: "\\u%04x", byte)
            buf.append(contentsOf: hex.utf8)
        default:
            buf.append(byte) // raw byte — preserves multi-byte UTF-8 intact
        }
    }
    buf.append(0x22) // closing "
    return buf
}

/// Build a complete JSON output message as raw Data, bypassing Swift String
/// to avoid UTF-8 validation that corrupts incomplete sequences at chunk boundaries.
func serializeOutputData(sessionId: String, data: Data) -> Data {
    var buf: [UInt8] = []
    buf.append(contentsOf: #"{"type":"output","session_id":"#.utf8)
    buf.append(contentsOf: jsonString(sessionId).utf8)
    buf.append(contentsOf: #","data":"#.utf8)
    buf.append(contentsOf: jsonEscapeBytes(data))
    buf.append(contentsOf: "}".utf8)
    return Data(buf)
}

/// Returns the number of trailing bytes that form an incomplete UTF-8 sequence.
func incompleteUTF8Tail(_ data: Data) -> Int {
    guard !data.isEmpty else { return 0 }
    for i in 1...min(3, data.count) {
        let byte = data[data.endIndex - i]
        if byte & 0x80 == 0 {
            return 0 // ASCII — sequence is complete
        } else if byte & 0xC0 != 0x80 {
            // Start byte — check if sequence is complete
            let expected: Int
            if byte & 0xE0 == 0xC0 { expected = 2 }
            else if byte & 0xF0 == 0xE0 { expected = 3 }
            else if byte & 0xF8 == 0xF0 { expected = 4 }
            else { return 0 } // Invalid start byte
            return i < expected ? i : 0
        }
        // Continuation byte (10xxxxxx) — keep scanning backward
    }
    return min(3, data.count) // All continuation bytes, no start byte
}
